### PR TITLE
Typescriptifying release step of functions deployment

### PR DESCRIPTION
**Pull request by joehan**
### Description
Refactoring release step of functions deploy to use Typescript. This PR doesn't have any significant functional changes - I wanted to first change this over to Typescript so that  it would be easier to understand what is actually being changed in future PRs.

### Scenarios Tested

Ran a series of function deploys to test the following situations:
- Create 1 function
- Create multiple function
- Update function
- Delete, create, update functions all in one deploy
- Create, update functions, N to prompt when asked to delete functions
- Create scheduled function
- Update scheduler function to different schedule
- Delete scheduled function

### CODE REVIEW

File: src/deploy/functions/release.ts [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+  logger.info(
+    "In a few minutes, you can check status at " + utils.consoleUrl(projectId, "/functions/logs")
+  );
+  logger.info(
+    "You can use the --only flag to deploy only a portion of your functions in the future."
+  );
+  deployments = []; // prevents analytics tracking of deployments
+}
+
+export async function release(context: any, options: any, payload: any): Promise<void> {
```
Comment (inlined) [inline src/deploy/functions/release.ts:133]: This is a 340LOC method that has nested conditionals since we can't early exit anywhere. It would probably be cleaner to actually use helper methods. Just a drive-by comment since this code was just translated, not written in the PR.

File: src/deploy/functions/release.ts [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+    if (!options.force) {
+      proceed = await promptOnce({
+        type: "confirm",
+        name: "confirm",
+        default: false,
+        message:
+          "Would you like to proceed with deletion? Selecting no will continue the rest of the deployments.",
+      });
+    }
+    if (proceed) {
```
Comment (inlined) [inline src/deploy/functions/release.ts:354]: nit: If we tested for `!proceed` (which I'd probably call `skip` to avoid a double negative) we would have the smaller block sooner. Then we wouldn't forget what the `else` block nearly 100 lines lower is referring to.

File: src/deploy/functions/release.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+      if (deployments.length !== 0) {
+        utils.logBullet(clc.bold.cyan("functions: ") + "continuing with other deployments.");
+      }
+    }
+  }
+  // filter out functions that are excluded via --only and --except flags
+  const functionsInDeploy = functionsInfo.filter((trigger: helper.CloudFunctionTrigger) => {
+    return functionFilterGroups.length > 0 ? _.includes(deleteReleaseNames, trigger.name) : true;
+  });
+  await createOrUpdateSchedulesAndTopics(
```
Comment (inlined) [inline src/deploy/functions/release.ts:433]: nit: the name for "createOrUpdate" is typically "upsert"

File: src/deploy/functions/release.ts [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+    context.projectId,
+    functionsInDeploy,
+    existingScheduledFunctions,
+    appEngineLocation
+  );
+  const allOps = await utils.promiseAllSettled(
+    _.map(deployments, async (op) => {
+      const res = await op.retryFunction();
```
Comment (inlined) [inline src/deploy/functions/release.ts:441]: If we only ever call this once, why is it called `retryFunction`?

**Review by inlined — APPROVED** [review]
Reviewed this one first to get into the groove of reviewing this codebase. I suspect this was largely just a rename as TS and changing a few syntax styles. I know you didn't write any of this whole cloth and don't expect any of my comments to actually be taken as action items.

For future reviews like this though, I like to do the fork as one commit and the edits as another. It offers the reviewer a way to scan through the changes much more easily.

File: src/deploy/functions/release.ts (hunk 1 above)
Comment (joehan) [reply src/deploy/functions/release.ts:133]: Agree 100 percent - this is getting torn to shreds in a coming PR
